### ISSUE TITLE
Faiss bad interactions with pytorch and transformers

### ISSUE THREAD

**Issue by pberger514** [facebookresearch/faiss#2850]
# Summary

Training `IndexIVFFlat` segfaults in the presence of transformers pytorch models.

# Platform

OS: macOS 13.3.1, Intel i7

Faiss version: 1.7.4

Installed from: anaconda, but then needed to symlink libmkl_intel 2 to 1. 

Running on:
- CPU

Interface: 
- Python

# Reproduction instructions
`conda install -c pytorch faiss-cpu`
`conda install transformers`

```
import numpy as np
import faiss
from transformers import AutoModel, AutoTokenizer


class SentenceTransformer:
    def __init__(self, model_name="all-MiniLM-L6-v2"):
        model_ckpt = f"sentence-transformers/{model_name}"
        self.tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
        self.model = AutoModel.from_pretrained(model_ckpt)

def test_faiss():
    st = SentenceTransformer()
    
    d = 64  # dimension
    nb = 100000  # database size
    nq = 10000  # nb of queries
    np.random.seed(1234)  # make reproducible
    xb = np.random.random((nb, d)).astype("float32")
    xb[:, 0] += np.arange(nb) / 1000.0
    xq = np.random.random((nq, d)).astype("float32")
    xq[:, 0] += np.arange(nq) / 1000.0

    nlist = 100
    k = 4
    quantizer = faiss.IndexFlatL2(d)  # the other index
    index = faiss.IndexIVFFlat(quantizer, d, nlist)

    assert not index.is_trained
    index.train(xb)
    assert index.is_trained

    index.add(xb)  # add may be a bit slower as well

    D, I = index.search(xq, k)  # actual search
    result_nprobe_1 = [
        [9900, 10500, 9831, 10808],
        [11055, 10812, 11321, 10260],
        [11353, 10164, 10719, 11013],
        [10571, 10203, 10793, 10952],
        [9582, 10304, 9622, 9229],
    ]
    assert np.sum(I[-5:] == result_nprobe_1) > 1

    index.nprobe = 10  # default nprobe is 1, try a few more
    D, I = index.search(xq, k)
    result_nprobe_10 = [
        [9900, 10500, 9309, 9831],
        [11055, 10895, 10812, 11321],
        [11353, 11103, 10164, 9787],
        [10571, 10664, 10632, 9638],
        [9628, 9554, 10036, 9582],
    ]
    assert np.all(I[-5:] == result_nprobe_10)

if __name__ == "__main__":
    test_faiss()
```

Leads to segfault.

**Comment by mdouze** [comment]
making a symlink is almost certainly a bad idea. What is the pytorch version?

**Comment by pberger514** [comment]
torch 1.13.1

I ran conda install but the lib was not found on import so required the symlink.

**Comment by algoriddle** [comment]
I changed the conda package to require mkl =2023 going forward both at build and runtime. Previously it was built with 2021, but defined mkl >= 2021 as its runtime requirement. I'm crossing fingers this fixes the problem. You can try this with the nightly build `conda install faiss-gpu -c pytorch/label/nightly`.
